merlin/models/tf/models/retrieval.py

```diff
@@ -300,3 +309,277 @@ def YoutubeDNNRetrievalModel(
+    Training with sampled_softmax is enabled by default [2]_ [3]_ [4]_.
+
+    Example Usage::
+        model = YoutubeDNNRetrievalModelV2(schema, num_sampled=100)
+        model.compile(optimizer="adam")
+        model.fit(train_data, epochs=10)
+
+    References
+    ----------
+    .. [1] Covington, Paul, Jay Adams, and Emre Sargin.
+        "Deep neural networks for youtube recommendations."
+        Proceedings of the 10th ACM conference on recommender systems. 2016.
+
+    .. [2] Yoshua Bengio and Jean-Sébastien Sénécal. 2003. Quick Training of Probabilistic
+       Neural Nets by Importance Sampling. In Proceedings of the conference on Artificial
+       Intelligence and Statistics (AISTATS).
+
+    .. [3] Y. Bengio and J. S. Senecal. 2008. Adaptive Importance Sampling to Accelerate
+       Training of a Neural Probabilistic Language Model. Trans. Neur. Netw. 19, 4 (April
+       2008), 713–722. https://doi.org/10.1109/TNN.2007.912312
+
+    .. [4] Jean, Sébastien, et al. "On using very large target vocabulary for neural
+        machine translation." arXiv preprint arXiv:1412.2007 (2014).
+
+
+    Parameters
+    ----------
+    schema: Schema
+        The `Schema` with the input features
+    candidate_id_tag : Tag
+        The tag to select cnadidate-id feature, by default `Tags.ITEM_ID`
+    top_block: tf.keras.layers.Layer
+        The hidden layers to apply on top of the features representation
+        vector.
+    inputs: tf.keras.layers.Layer, optional
+        The input layer to encode input features (sparse and context features)
+        If not specified, the input layer is inferred from the schema
+        By default None
+    post: Optional[tf.keras.layers.Layer], optional
+        The optional layer to apply on top of the query encoder.
+    logits_temperature: float, optional
+        Parameter used to reduce model overconfidence, so that logits / T.
+        Defaults to 1.
+    num_sampled: int, optional
+        When sampled_softmax is enabled, specify the number of
+        negative candidates to generate for each batch.
+        By default 100
+    min_sampled_id: int, optional
+        The minimum id value to be sampled with sampled softmax.
+        Useful to ignore the first categorical
+        encoded ids, which are usually reserved for <nulls>,
+        out-of-vocabulary or padding.
+        By default 0.
+    """
+    if not inputs:
+        inputs = schema
+
+    candidate = schema.select_by_tag(candidate_id_tag)
+    if not candidate:
+        raise ValueError(f"The schema should contain a feature tagged as `{candidate_id_tag}`")
+    candidate = candidate.first
+    num_classes = candidate.int_domain.max + 1
+
+    query = Encoder(inputs, top_block, post=post)
+
+    if not outputs:
+        warnings.warn(
+            "By default, the YoutubeDNN model is trained using popularity-bases sampled softmax."
+        )
+        candidate_table = query.first["categorical"][candidate.name]
```

Comment:
I understand that `.first()` here is a default keras method for (Encoder) Model that returns the first layer (block) of the model. There can be a case where the InputBlock wont' be the first layer?
Some scenarios potential scenarios for that could be:
- Adding a pre to InputBlock
- Providing a pre for the model.fit(), which is currently used for `SequenceMaskRandom`

Would `model.get_blocks_by_name()` be a safer way to retrieve the input block?

Reply:
Here first is a property of the Encoder class that returns the first block using `self.blocks[0]`. The blocks of the encoder are defined in this [line](https://github.com/NVIDIA-Merlin/models/blob/637c5691820312d6f82d891c742ae3a81a7b6db8/merlin/models/tf/core/encoder.py#L54).  Based on this definition,  I believe `self.blocks[0]` will always be related to the whole input_block (including the `pre` defined inside the input block)

Reply:
ok. Makes sense to me @sararb